src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -23,50 +26,153 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
+
+        public DefaultRazorSemanticTokensInfoService(IClientLanguageServer languageServer,
+            RazorDocumentMappingService documentMappingService)
         {
-            return GetSemanticTokens(codeDocument, range: null);
+            _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
         }
 
-        public override SemanticTokens GetSemanticTokens(RazorCodeDocument codeDocument, Range range)
+        public override Task<SemanticTokens> GetSemanticTokens(RazorCodeDocument codeDocument, TextDocumentIdentifier textDocumentIdentifier, CancellationToken cancellationToken)
+        {
+            return GetSemanticTokens(codeDocument, textDocumentIdentifier, range: null, cancellationToken);
+        }
+
+        public override async Task<SemanticTokens> GetSemanticTokens(
+            RazorCodeDocument codeDocument,
+            TextDocumentIdentifier textDocumentIdentifier,
+            Range range,
+            CancellationToken cancellationToken)
         {
             if (codeDocument is null)
             {
                 throw new ArgumentNullException(nameof(codeDocument));
             }
 
-            var semanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument, range);
+            var razorSemanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument, range);
+            var cSharpSemanticRanges = await GetCSharpSemanticRanges(codeDocument, textDocumentIdentifier, range, cancellationToken);
+            var combinedSemanticRanges = CombineSemanticRanges(razorSemanticRanges, cSharpSemanticRanges);
 
-            var semanticTokens = ConvertSemanticRangesToSemanticTokens(semanticRanges, codeDocument);
+            var razorSemanticTokens = ConvertSemanticRangesToSemanticTokens(combinedSemanticRanges, codeDocument);
 
-            return semanticTokens;
+            return razorSemanticTokens;
         }
 
-        public override SemanticTokensFullOrDelta GetSemanticTokensEdits(RazorCodeDocument codeDocument, string previousResultId)
+        public override async Task<SemanticTokensFullOrDelta> GetSemanticTokensEdits(
+            RazorCodeDocument codeDocument,
+            TextDocumentIdentifier textDocumentIdentifier,
+            string previousResultId,
+            CancellationToken cancellationToken)
         {
             if (codeDocument is null)
             {
                 throw new ArgumentNullException(nameof(codeDocument));
             }
 
-            var semanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument);
+            var razorSemanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument);
+            var cSharpSemanticRanges = await GetCSharpSemanticRanges(codeDocument, textDocumentIdentifier, range: null, cancellationToken);
```

Comment:
So does this mean for every edit we're going to re-query the entire C# documents semantic tokens? Also since we're providing them an empty "previousResultId" do they do any work to cache their responses or are they doing fresh calculations each time?

Also given the fact that this is queried multiple times a second it's concerning that we'd be hamming the C# semantic token service without cacheability that frequently.